route_guide/route_guide_client.cc

```diff
@@ -23,238 +23,346 @@
+                           / kCoordFactor_
+                             << ", "
+                             << data.feature.location().longitude()
+                           / kCoordFactor_
+                             << std::endl;
+                 } else {
+                   std::cout << "Found feature called "
+                             << data.feature.name() << " at "
+                             << data.feature.location().latitude()
+                           / kCoordFactor_
+                             << ", "
+                             << data.feature.location().longitude()
+                           / kCoordFactor_
+                             << std::endl;
+                 }
+               })
+               // .finished callback is invoked when the call is finished.
+               // In this callback we just start our continuation.
+               .finished([](
+                             auto&& data,
+                             auto& continuation,
+                             auto&& resp) {
+                 continuation.Start(grpc::Status::OK);
+               }));
+  }
 
-    Status status = done.Wait();
-    if (status.ok()) {
-      std::cout << "ListFeatures rpc succeeded." << std::endl;
-    } else {
-      std::cout << "ListFeatures rpc failed." << std::endl;
-    }
+  // The `ListFeatures()` function demonstrates a streaming RPC. It is
+  // comparable to `GetFeature()`, but instead of sending a single `Point`
+  // the client sends a `Rectangle`. The server responds with a stream of
+  // messages containing addresses that fall inside the Rectangle's area.
+  auto ListFeatures() {
+    struct Data {
+      Rectangle rectangle; // Request from the client.
+      Feature feature; // Response from the server.
+    };
+    return client_->Call<RouteGuide, Rectangle, Stream<Feature>>(
+               "ListFeatures")
+        | Client::Handler()
+              .context(Data{Rectangle{}, Feature{}})
+              // Prepare the rectangle area and send to the server as a
+              // request.
+              .ready([](auto&& data, auto& call) {
+                data.rectangle.mutable_lo()->set_latitude(400000000);
+                data.rectangle.mutable_lo()->set_longitude(-750000000);
+                data.rectangle.mutable_hi()->set_latitude(420000000);
+                data.rectangle.mutable_hi()->set_longitude(-730000000);
+                std::cout << "Looking for features between 40, -75"
+                          << " and 42, -73" << std::endl;
+                call.WriteLast(data.rectangle);
+              })
+              // The .body() callback is invoked every time the server
+              // sends a response. The response contains the addresses
+              // that fall within the requested rectangle's area.
+              .body([this](auto&& data, auto& call, auto&& feature) {
+                // Check if the response from the server is valid.
+                if (!feature) {
```

Comment:
As with the comment below about type-safe checks:

```suggestion
                if (feature == nullptr) {
```